go/libraries/doltcore/doltdb/table_edit_session.go

```diff
@@ -0,0 +1,239 @@
+	tables     map[string]*SessionedTableEditor
+	writeMutex *sync.RWMutex // This mutex is specifically for changes that affect the TES or all STEs
+}
+
+// TableEditSessionProps are properties that define different functionality for the TableEditSession.
+type TableEditSessionProps struct {
+	ForeignKeyChecksDisabled bool // If disabled, then all foreign key checks and updates are ignored
+}
+
+// CreateTableEditSession creates and returns a TableEditSession. Inserting a nil root is not an error, as there are
+// locations that do not have a root at the time of this call. However, a root must be set through SetRoot before any
+// table editors are returned.
+func CreateTableEditSession(root *RootValue, props TableEditSessionProps) *TableEditSession {
+	return &TableEditSession{
+		Props:      props,
+		root:       root,
+		tables:     make(map[string]*SessionedTableEditor),
+		writeMutex: &sync.RWMutex{},
+	}
+}
+
+// GetTableEditor returns a SessionedTableEditor for the given table. If a schema is provided and it does not match the one
+// that is used for currently open editors (if any), then those editors will reload the table from the root.
+func (tes *TableEditSession) GetTableEditor(ctx context.Context, tableName string, tableSch schema.Schema) (*SessionedTableEditor, error) {
+	tes.writeMutex.Lock()
+	defer tes.writeMutex.Unlock()
+
+	return tes.getTableEditor(ctx, tableName, tableSch)
+}
+
+// GetRoot returns an updated root with all of the changed tables.
+func (tes *TableEditSession) GetRoot(ctx context.Context) (*RootValue, error) {
+	tes.writeMutex.Lock()
+	defer tes.writeMutex.Unlock()
+
+	rootMutex := &sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	wg.Add(len(tes.tables))
+
+	newRoot := tes.root
+	var tableErr error
+	var rootErr error
+	for tableName, ste := range tes.tables {
+		// we can run all of the Table calls concurrently as long as we guard updating the root
+		go func(tableName string, ste *SessionedTableEditor) {
+			defer wg.Done()
+			updatedTable, err := ste.tableEditor.Table()
+			// we lock immediately after doing the operation, since both error setting and root updating are guarded
+			rootMutex.Lock()
+			defer rootMutex.Unlock()
+			if err != nil {
+				if tableErr == nil {
+					tableErr = err
+				}
+				return
+			}
+			newRoot, err = newRoot.PutTable(ctx, tableName, updatedTable)
+			if err != nil && rootErr == nil {
+				rootErr = err
+			}
+		}(tableName, ste)
+	}
+	wg.Wait()
+	if tableErr != nil {
+		return nil, tableErr
+	}
+	if rootErr != nil {
+		return nil, rootErr
+	}
+
+	tes.root = newRoot
+	return newRoot, nil
+}
+
+// SetRoot uses the given root to set all open table editors to the state as represented in the root. If any
+// tables are removed in the root, but have open table editors, then the references to those are removed. If those
+// removed table's editors are used after this, then the behavior is undefined.
+func (tes *TableEditSession) SetRoot(ctx context.Context, root *RootValue) error {
```

Comment:
This is explicitly and intentionally not an error. There is only one time that a table editor should ever be deleted, and that's when the underlying table is deleted.

So updating the root will update all open table editors, that is the intent of this function. `TableEditSession` only handles updating row data of a table, and thus any other changes to a table must be done externally. Index updates, schema changes, foreign key changes...whatever it is, those happen outside of the editor. So in order for editors to pick up on those changes, you'd either have to 1) close all editors, make the changes, and require each previously-open editor to fetch a new one, or 2) block all writes session-wide and edit each editor in-place. I've gone with the 2nd, which fits with the _I only worry about my table_ design goal.

The only possible issue in my current implementation is if a caller sets the root without first grabbing the updated root from an edit session, as that may cause you to lose changes. I thought about adding a function with a callback to make this a bit more explicit (and prevent this issue from occurring), so I can add that if you feel it's clearer in its intent.

Reply:
I would just make the comment here a little more explicit about how this is expected to be used (for schema changes) and add some caveats. Like, if I update a table's rows outside this edit session, and update it in the root that I pass here, and there is a table editor open for that table, then the edits that I made outside the session get lost. What I'm getting at is that there are many different root values you could pass this thing that would lead to bad / undefined behavior, so you should attempt to document them and give guidance on what to do instead.